man/crass.1
.\"Modified from man(1) of FreeBSD, the NetBSD mdoc.template, and mdoc.samples.
.\"See Also:
.\"man mdoc.samples for a complete listing of options
.\"man mdoc for the short list of editing options
.\"/usr/share/misc/mdoc.template
.Dd 5/09/16                
.Dt crass 1       
.Os Darwin
.Sh NAME                 
.Nm crass
.Nd the CRISPR Assembler.
.Sh SYNOPSIS             
.Nm
.Op Fl abcdDefgGhkKlLnorsSVw
.Ar

.Sh DESCRIPTION         
.Nm
is a tool for finding and assembling reads from genomic and metagenomic datasets that contain 
.Em C Ns lustered
.Em R Ns egularly
.Em I Ns nterspersed
.Em S Ns hort
.Em P Ns alindromic
.Em R Ns epeats
(CRISPR).  
.Nm
searches through the dataset and identifies reads which contain repeated K-mers that are of a specific length and are 
separated by a spacer sequence.  These possible direct repeats are then curated internally to remove bad matches and 
then reads containing direct repeats are then outputed for further analysis.  

.Pp
.Sh OPTIONS

.Bl -tag -width -indent
.It  
.Nm
.Op Fl eghrGL
.Op Fl a Ar LAYOUT_TYPE
.Op Fl b Ar INT
.Op Fl c Ar COLOUR_TYPE
.Op Fl d Ar INT
.Op Fl f Ar INT
.Op Fl k Ar INT
.Op Fl K Ar INT
.Op Fl l Ar INT
.Op Fl n Ar INT
.Op Fl o Ar DIR
.Op Fl s Ar INT
.Op Fl w Ar INT
.Op Fl D Ar INT
.Op Fl K Ar INT
.Op Fl S Ar INT
.Ar

.It Fl a Ar LAYOUT_TYPE Fl "\^\-layoutAlgorithm" Ar LAYOUT_TYPE
The Graphviz layout algorithm to be used when rendering graphs.
.It Fl b Ar INT Fl "\^\-numBins" Ar INT
The number of colour bins for the output graph. Default is to have as many colours as there are different values for the coverage of Nodes in the graph.
.It Fl c Ar COLOUR_TYPE Fl "\^\-graphColour" Ar COLOUR_TYPE
The colour scheme for the output graph based on the coverage of each spacer in the CRISPR, can be one from:
.Bl -tag -width -indent
.It red-blue
Low coverage spacers are coloured in red and high coverage spaceres are coloured blue. Intermediates are coloured in shades of purple. 
.It blue-red
Low coverage spacers are coloured in blue and high coverage spaceres are coloured red. Intermediates are coloured in shades of purple.
.It green-red-blue 
Three tone colouring with low coverage spacers in green and high coverage spacers in blue.
.It red-blue-green
Three tone colouring with low coverage spacers in blue and high coverage spacers in green.
.El
.It Fl d Ar INT Fl "\^\-minDR" Ar INT             
The minimim length of the direct repeat to search for [Default: 23] 
.It Fl D Ar INT Fl "\^\-maxDR" Ar INT             
The Maximum length of the direct repeat to search for [Default: 47] 
.It Fl e Ar "" Fl "\^\-noDebugGraph"
Option available only when DEBUG preoprocessor symbol is set. Will turn off generating debugging graphs
.It Fl f Ar INT  Fl "\^\-covCutoff" Ar INT           
Defines the minimim number of spacers that a putative CRISPR must contain to be considered real. [Default: 3]
.It Fl g Ar "" Fl "\^\-logToScreen"
Print the logging info to stdout rather than to a file
.It Fl G Ar ""  Fl "\^\-showSingletons" Ar ""
Set to show unattached spacers in the graph output  
.It Fl h Ar ""  Fl "\^\-help" Ar ""           
Output basic usage informtion to screen
.It Fl l Ar INT Fl "\^\-logLevel" Ar INT
The level of verbosity to ouput in the
.Nm 
log file 
.It Fl k Ar INT Fl "\^\-kmerCount" Ar INT            
The number of kmers at two direct repeats must share to be considered part of the same cluster [Default: 12]
.It Fl K Ar INT Fl "\^\-graphNodeLen" Ar INT            
The length of the kmer used to define a node in the graph.  The lower the number the more connected the graph will be but also increases the chance of false positive edges [Default: 7]
.It Fl n Ar INT Fl "\^\-minNumRepeats" Ar INT            
The minimim number of repeats that a candidate CRISPR locus must contain to be considered 'real' [Default: 2]
.It Fl o Ar LOCATION  Fl "\^\-outDir" Ar LOCATION          
The name of the ouput directory for the output files [Default: ./]
.It Fl r Ar "" Fl "\^\-noRendering" Ar ""
Option only available when the '--enable-rendering' configure option is set.  Will turn off the generation of image files.
.It Fl s Ar INT Fl "\^\-minSpacer" Ar INT            
The minimim length of the spacer to search for [Default: 26]
.It Fl S Ar INT Fl "\^\-maxSpacer" Ar INT          
The maximim length of the spacer to search for [Default: 50]
.It Fl V   Ar ""  Fl "\^\-version" Ar ""        
Print version and copy right information
.It Fl w Ar INT Fl "\^\-windowLength" Ar INT            
The length of the window size for searching a genome.  Must be between 6 - 9 [Default: 8]
.El

.\" .Sh ENVIRONMENT      \" May not be needed
.\" .Bl -tag -width "ENV_VAR_1" -indent \" ENV_VAR_1 is width of the string ENV_VAR_1
.\" .It Ev ENV_VAR_1
.\" Description of ENV_VAR_1
.\" .It Ev ENV_VAR_2
.\" Description of ENV_VAR_2
.\" .El                      
.Sh FILES                
.Bl -tag -width -indent
.It Pa crass.<TIMESTAP>.log
Log file containing information about the last execution of 
.Nm
.It Pa Group_<NUM>_<DNA>.fa
Fasta file of all reads from a DR type.  
.It Pa Spacers_<NUM>_<DNA>.spacers.gv
File representing the graph of the DR type in Graphviz format
.It Pa crass.<TIMESTAMP>.keys.gv
A file in graphviz format that contains all of the colour codes for the coverage values in the output graph
.It Pa crass.crispr
A crispr file representing all the information about each of the DR types identified
.El  
.Sh DIAGNOSTICS       \" May not be needed
.Ex -std 

.Sh SEE ALSO 
.Xr crass-assembler 1 , 
.Xr dot 1 ,
.Xr crisprtools 1 
.\".Xr b 2 ,
.\".Xr a 3 ,
.\".Xr b 3 
.\" .Sh BUGS              \" Document known, unremedied bugs 
.\" .Sh HISTORY           \" Document history if command behaves in a unique manner


ChangeLog
commit 54af4fb2cb1134de1685cdb3dfbee177829ca300
Author: ctSkennerton <c.skennerton@gmail.com>
Date:   Wed Apr 17 10:27:15 2013 +1000

    Fixed bug in cap3 assembly wrapper
    The path to the input file wasn't getting passed through properly
    causing cap3 to both fail and produce a number of hidden files.
    This fix properly passes the input file to cap3.

commit 4b5487710626d638d4197842699bb3e53a99d9d8
Author: ctSkennerton <c.skennerton@gmail.com>
Date:   Wed Apr 17 09:54:40 2013 +1000

    Fixed velvet wrapper bug
    If the -o option was not set on the command line the velvet wrapper
    would fail.  Now added a default setting of the current working
    directory so that it does not fail

commit 08a904e784038e69f365d5e0747253c34008024a
Author: ctSkennerton <c.skennerton@gmail.com>
Date:   Wed Apr 17 09:41:41 2013 +1000

    Fixed bug with crass-assembler
    A throwback from the previous interface was that the first
    argument to the program was getting shifted off and not being
    processed by getopts, which was causing all sorts of problems

commit c7ac0b37f686dfbe31b32e3a3a0def1424e5a1a5
Merge: f065f54 970477f
Author: ctSkennerton <c.skennerton@gmail.com>
Date:   Mon Dec 24 15:20:05 2012 +1000

    Merge branch 'master' of github.com:ctSkennerton/crass

commit 970477fc366bc47927f5a05876c65d92047aee06
Author: ctSkennerton <c.skennerton@gmail.com>
Date:   Wed Oct 3 22:18:51 2012 +1000

    forgot to update the version number in the manual

commit b6e4a270b9d637776f84c0b80f22456afe8ff600
Author: ctSkennerton <c.skennerton@gmail.com>
Date:   Wed Oct 3 19:24:26 2012 +1000

    Updated the version number to 0.3.1 in configure.ac

commit 8ac2e65aa1b7b23cb08040d5df96ffe21b2b1d6d
Author: ctSkennerton <c.skennerton@gmail.com>
Date:   Wed Oct 3 19:08:32 2012 +1000

    Fixed a bug in ReadHolder::dropPartials()
    The logic for dropping an end partial repeat was incorrect such
    that it was removing direct repeats that were close to but not
    hanging over the edge of the read.  This had flow on effects in
    libcrispr#longReadFinder() where the last spacer in the read was
    getting removed when it should not have been

commit 287e194e25c656d21a3acf2f77a39e984a92d69b
Author: ctSkennerton <c.skennerton@gmail.com>
Date:   Tue Oct 2 16:00:30 2012 +1000

    changed the default for the high DR size to 47 (was 45)

commit d30e5308bd26dea7d42affa88c0ebb5fc8ed6f9b
Author: ctSkennerton <c.skennerton@gmail.com>
Date:   Tue Oct 2 15:49:43 2012 +1000

    Fixed the function call to NodeManager::dumpReads()
    During the merging of branches the call to NodeManager::dumpReads()
    became out of sync with the method definition.  This change removes
    one of the parameters in WorkHorse::dumpReads() to get the two
    functions back in sync

commit 8b44ba32f85b3d0c170cfd29478b6345a8903d39
Author: ctSkennerton <c.skennerton@gmail.com>
Date:   Tue Oct 2 15:30:43 2012 +1000

    Removed an errant print statement

commit e687f4af5a508a10a91415261265e3d4df158095
Author: ctSkennerton <c.skennerton@gmail.com>
Date:   Tue Oct 2 15:26:29 2012 +1000

    Fixed a bug in libcrispr#longReadFinder()
    The if statement controlling the number of repeats in a read had
    a strictly greater than comparison rather than the intended inclusive
    operator.  This meant that by default the number of repeats in a read
    had to be 4 not 3 which was intended

commit e1028e5dce206027d58aff801eb6bf6f6ceaca60
Merge: 0847fb3 d7e32c1
Author: ctSkennerton <c.skennerton@gmail.com>
Date:   Sun Sep 30 22:28:30 2012 +1000

    Merge branch 'dr_aligner'
    
    Conflicts:
    	doc/manual.tex
    	src/crass/Makefile.am
    	src/crass/WorkHorse.cpp

commit d7e32c124a4cdb41a1455bd44c6a8803b930efd4
Author: ctSkennerton <c.skennerton@gmail.com>
Date:   Sun Sep 30 22:14:35 2012 +1000

    Updated to version 0.3.0
    This version implements a number of changes, both 'under-the-hood'
    and visible to the user.  First, Crass now has both short and long form
    for each of it's options.  Second, the assembly wrapper

src/crass/crass.cpp
/*
 *  crass.cpp is part of the CRisprASSembler project
 *  
 *  Created by Connor Skennerton.
 *  Copyright 2011 Connor Skennerton & Michael Imelfort. All rights reserved. 
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 *
 *
 *                     A B R A K A D A B R A
 *                      A B R A K A D A B R
 *                       A B R A K A D A B
 *                        A B R A K A D A       	
 *                         A B R A K A D
 *                          A B R A K A
 *                           A B R A K
 *                            A B R A
 *                             A B R
 *                              A B
 *                               A
 */

// system includes
#include <stdio.h>
#include <stdlib.h>
#include <limits.h>
#include <string>
#include <unistd.h>
#include <fcntl.h>
#include <istream>
#include <fstream>
#include <iostream>
#include <sstream>
#include <map>
#include <time.h>
#include <sys/types.h>
#include <sys/stat.h>

// local includes
#include "config.h"
#include "crass.h"
#include "crassDefines.h"
#include "LoggerSimp.h"
#include "WorkHorse.h"
#include "Rainbow.h"
#include "StlExt.h"
#include "Exception.h"
#include "SeqUtils.h"


//**************************************
// user input + system
//**************************************

void usage(void) 
{
    std::cout<<"Compiler Options:"<<std::endl;
    std::cout<<"RENDERING =";
#ifdef RENDERING
    std::cout<<" 1"<<std::endl;
#else
    std::cout<<" 0"<<std::endl;
#endif
    std::cout<<"DEBUG =";
#ifdef DEBUG
    std::cout<<" 1"<<std::endl;
#else
    std::cout<<" 0"<<std::endl;
#endif
    std::cout<<"MEMCHECK =";
#ifdef MEMCHECK
    std::cout<<" 1"<<std::endl;
#else
    std::cout<<" 0"<<std::endl;
#endif
    std::cout<<"ASSEMBER =";
#ifdef PERFORM_CRASS_ASSEMBLY
    std::cout<<" 1"<<std::endl;
#else
    std::cout<<" 0"<<std::endl;
#endif
    std::cout<<"VERBOSE_LOGGER =";
#ifdef SUPER_LOGGING
    std::cout<<" 1"<<std::endl;
#else
    std::cout<<" 0"<<std::endl;
#endif
    std::cout<<"Search Debugger = ";
#ifdef SEARCH_SINGLETON
    std::cout<<" 1"<<std::endl;
#else
    std::cout<<" 0"<<std::endl;
#endif
    std::cout<<std::endl;
    std::cout<<"Usage:  "<<PACKAGE_NAME<<"  [options] { inputFile ...}"<<std::endl<<std::endl;
    std::cout<<"General Options:"<<std::endl;
    std::cout<< "-h --help                    This help message"<<std::endl;
    std::cout<< "-l --logLevel        <INT>   Output a log file and set a log level [1 - "<<CRASS_DEF_MAX_LOGGING<<"]"<<std::endl;
    std::cout<< "-o --outDir          <DIR>   Output directory [default: .]"<<std::endl;
    std::cout<< "-V --version                 Program and version information"<<std::endl;
    std::cout<< "-g --logToScreen             Print the logging information to screen rather than a file"<<std::endl;
    std::cout<<std::endl;
    std::cout<<"CRISPR Identification Options:"<<std::endl;
    std::cout<< "-d --minDR           <INT>   Minimim length of the direct repeat"<<std::endl; 
    std::cout<< "                             to search for [Default: "<<CRASS_DEF_MIN_DR_SIZE<<"]"<<std::endl;
    std::cout<< "-D --maxDR           <INT>   Maximim length of the direct repeat"<<std::endl; 
    std::cout<< "                             to search for [Default: "<<CRASS_DEF_MAX_DR_SIZE<<"]"<<std::endl;
    std::cout<< "-n --minNumRepeats   <INT>   Total number of direct repeats in a CRISPR for"<<std::endl;
    std::cout<< "                             it to be considered real [Default: "<<CRASS_DEF_DEFAULT_MIN_NUM_REPEATS<<"]"<<std::endl;
    std::cout<< "-s --minSpacer       <INT>   Minimim length of the spacer to search for [Default: "<<CRASS_DEF_MIN_SPACER_SIZE<<"]"<<std::endl;
    std::cout<< "-S --maxSpacer       <INT>   Maximim length of the spacer to search for [Default: "<<CRASS_DEF_MAX_SPACER_SIZE<<"]"<<std::endl;
    std::cout<< "-w --windowLength    <INT>   The length of the search window. Can only be"<<std::endl; 
    std::cout<< "                             a number between "<<CRASS_DEF_MIN_SEARCH_WINDOW_LENGTH<<" - "<<CRASS_DEF_MAX_SEARCH_WINDOW_LENGTH<<" [Default: "<<CRASS_DEF_OPTIMAL_SEARCH_WINDOW_LENGTH<<"]"<<std::endl;
    /*std::cout<< "-x --spacerScalling  <REAL>  A decimal number that represents the reduction in size of the spacer"<<std::endl;
    std::cout<< "                             when the --removeHomopolymers option is set [Default: "<<CRASS_DEF_HOMOPOLYMER_SCALLING<<"]"<<std::endl;
    std::cout<< "-y --repeatScalling  <REAL>  A decimal number that represents the reduction in size of the direct repeat"<<std::endl;
    std::cout<< "                             when the --removeHomopolymers option is set [Default: "<<CRASS_DEF_HOMOPOLYMER_SCALLING<<"]"<<std::endl;
    std::cout<< "-z --noScalling              Use the given spacer and direct repeat ranges when --removeHomopolymers is set. "<<std::endl;
    std::cout<< "                             The default is to scale the numbers by "<<CRASS_DEF_HOMOPOLYMER_SCALLING<<" or by values set using -x or -y"<<std::endl;
    std::cout<< "-H --removeHomopolymers      Correct for homopolymer errors [default: no correction]"<<std::endl;
    std::cout<<std::endl;*/
    std::cout<<"CRISPR Assembly Options:"<<std::endl;
    std::cout<< "-f --covCutoff       <INT>   Remove groups with less than x spacers [Default: "<<CRASS_DEF_COVCUTOFF<<"]"<<std::endl;
    std::cout<< "-k --kmerCount       <INT>   The number of the kmers that need to be"<<std::endl; 
    std::cout<< "                             shared for clustering [Default: "<<CRASS_DEF_K_CLUST_MIN<<"]"<<std::endl;
    std::cout<< "-K --graphNodeLen    <INT>   Length of the kmers used to make crispr nodes [Default: "<<CRASS_DEF_NODE_KMER_SIZE<<"]"<<std::endl;
    std::cout<<std::endl;
    std::cout<<"Output Options: "<<std::endl;
#ifdef RENDERING
    std::cout<<"-a --layoutAlgorithm  <TYPE>  Graphviz layout algorithm to use for printing spacer graphs. The following are available:"<<std::endl;
    #ifdef HAVE_NEATO
    std::cout<<"                              \tneato"<<std::endl;
    #endif
    #ifdef HAVE_DOT
    std::cout<<"                              \tdot [Default]"<<std::endl;
    #endif
    #ifdef HAVE_FDP
    std::cout<<"                              \tfdp"<<std::endl;
    #endif
    #ifdef HAVE_SFDP
    std::cout<<"                              \tsfdp"<<std::endl;
    #endif
    #ifdef HAVE_CIRCO
    std::cout<<"                              \tcirco"<<std::endl;
    #endif
    #ifdef HAVE_TWOPI
    std::cout<<"                              \ttwopi"<<std::endl;
    #endif
#endif 
    std::cout<<"-b --numBins          <INT>   The number of colour bins for the output graph."<<std::endl;
    std::cout<<"                              Default is to have as many colours as there are"<<std::endl;
    std::cout<<"                              different values for the coverage of Nodes in the graph"<<std::endl;
    std::cout<<"-c --graphColour     <TYPE>   Defines the range of colours to use for the output graph"<<std::endl;
    std::cout<<"                              the different types available are:"<<std::endl;
    std::cout<<"                              red-blue, blue-red, green-red-blue, red-blue-green"<<std::endl;
    std::cout<<"-L --longDescription          Set if you want the spacer sequence printed along with the ID in the spacer graph. [Default: false]"<<std::endl;
    std::cout<<"-G --showSingltons            Set if you want to print singleton spacers in the spacer graph [Default: false]"<<std::endl;
#ifdef DEBUG
    std::cout<<"-e --noDebugGraph             Stops creation of debug .gv files even if the DEBUG preprocessor macro is set [Default: false]"<<std::endl;
#endif
#ifdef RENDERING
    std::cout<<"-r --noRendering              Stops rendering of .gv files even if the RENDERING preprocessor macro is set [Default: false]"<<std::endl;
#endif
#ifdef SEARCH_SINGLETON
    std::cout<<"--searchChecker       <FILE>  A file containing read headers that should be tracked through "<<PACKAGE_NAME<<std::endl;
#endif
}

void versionInfo(void) 
{
    std::cout<<std::endl<<PACKAGE_FULL_NAME<<" ("<<PACKAGE_NAME<<")"<<std::endl<<"version "<<PACKAGE_MAJOR_VERSION<<" subversion "<<PACKAGE_MINOR_VERSION<<" revison "<<PACKAGE_REVISION<<" ("<<PACKAGE_VERSION<<")"<<std::endl<<std::endl;
    std::cout<<"---------------------------------------------------------------"<<std::endl;
    std::cout<<"Copyright (C) 2011-2015 Connor Skennerton & Michael Imelfort"<<std::endl;
    std::cout<<"Copyright (C) 2016      Connor Skennerton"<<std::endl;
    std::cout<<"This program comes with ABSOLUTELY NO WARRANTY"<<std::endl;
    std::cout<<"This is free software, and you are welcome to redistribute it"<<std::endl;
    std::cout<<"under certain conditions: See the source for more details"<<std::endl;
    std::cout<<"---------------------------------------------------------------"<<std::endl;
}


int processOptions(int argc, char *argv[], options *opts) 
{
    int c;
    int index;
    while( (c = getopt_long(argc, argv, "a:b:c:d:D:ef:gGhk:K:l:Ln:o:rs:S:Vw:", long_options, &index)) != -1 ) 
    {
        switch(c) 
        {
            case 'a':
#if RENDERING                
                if (strcmp(optarg, "dot") && strcmp(optarg, "neato") && strcmp(optarg, "fdp") && strcmp(optarg, "sfdp") && strcmp(optarg, "twopi") && strcmp(optarg, "circo")) {
                    std::cerr<<PACKAGE_NAME<<" [WARNING]: '"<<optarg<<"' is not a recognised layout algorithm. Please choose from the following:"<<std::endl;
#ifdef HAVE_NEATO
                    std::cerr<<"\tneato"<<std::endl;
#endif
#ifdef HAVE_DOT
                    std::cerr<<"\tdot"<<std::endl;
#endif
#ifdef HAVE_FDP
                    std::cerr<<"\tfdp"<<std::endl;
#endif
#ifdef HAVE_SFDP
                    std::cerr<<"\tsfdp"<<std::endl;
#endif
#ifdef HAVE_CIRCO
                    std::cerr<<"\tcirco"<<std::endl;
#endif
#ifdef HAVE_TWOPI
                    std::cerr<<"\ttwopi"<<std::endl;
#endif
                    
                } else {
                    opts->layoutAlgorithm = optarg;
                }
#else 
                std::cerr<<PACKAGE_NAME<<" [WARNING]: Not compiled with the correct options to allow rendering of graphs.\nEither use the --enable-rendering option during ./configure and make sure that the graphviz layout programs are in your PATH"<<std::endl;
#endif
                break;
            case 'b': 
                from_string<int>(opts->coverageBins, optarg, std::dec);
                if (opts->coverageBins < 1) 
                {
                    std::cerr<<PACKAGE_NAME<<" [ERROR]: The number of bins for coverage cannot be less than 1"<<std::endl;
                    usage();
                    exit(1);
                }
                break;
            case 'c': 
                if (strcmp(optarg, "red-blue") == 0) 
                {
                    opts->graphColourType = RED_BLUE;
                } 
                else if(strcmp(optarg, "read-blue-green") == 0)
                {
                    opts->graphColourType = RED_BLUE_GREEN;
                }
                else if(strcmp(optarg, "blue-red") == 0)
                {
                    opts->graphColourType = BLUE_RED; 
                }
                else if(strcmp(optarg, "green-blue-red") == 0)
                {
                    opts->graphColourType = GREEN_BLUE_RED;
                }
                else
                {
                    std::cerr<<PACKAGE_NAME<<" [WARNING]: Unknown graph colour type "<<optarg<<" changing to default colour type (red-blue)"<<std::endl;
                    opts->graphColourType = RED_BLUE;
                }
                break;
            case 'd': 
                from_string<unsigned int>(opts->lowDRsize, optarg, std::dec);
                if (opts->lowDRsize < 8) 
                {
                    std::cerr<<PACKAGE_NAME<<" [WARNING]: The lower bound for direct repeat sizes cannot be "<<opts->lowDRsize<<" changing to "<<CRASS_DEF_MIN_DR_SIZE<<std::endl;
                    opts->lowDRsize = CRASS_DEF_MIN_DR_SIZE;
                }
                break;
            case 'D': 
                from_string<unsigned int>(opts->highDRsize, optarg, std::dec);
                break;
            case 'e':
#ifdef DEBUG
                opts->noDebugGraph = true;
#endif
                break;
            case 'f':
                from_string<int>(opts->covCutoff, optarg, std::dec);
                break;
            case 'g':
                 opts->logToScreen = true;
                break;
            case 'G':
            	opts->showSingles= true;
                break;
            case 'h': 
                versionInfo(); 
                usage();
                exit(0); 
                break;
            case 'k': 
                from_string<int>(opts->kmer_clust_size, optarg, std::dec);
                if (opts->kmer_clust_size < 4) 
                {
                    std::cerr<<PACKAGE_NAME<<" [WARNING]: Minimum value for kmer clustering size is: "<<4<<" changing to "<<CRASS_DEF_K_CLUST_MIN<<std::endl;
                    opts->kmer_clust_size = CRASS_DEF_K_CLUST_MIN;
                }
                break;
            case 'K': 
                from_string<int>(opts->cNodeKmerLength, optarg, std::dec);
                break;
            case 'l': 
                from_string<int>(opts->logLevel, optarg, std::dec);
                if(opts->logLevel > CRASS_DEF_MAX_LOGGING)
                {
                    std::cerr<<PACKAGE_NAME<<" [WARNING]: Specified log level higher than max. Changing log level to "<<CRASS_DEF_MAX_LOGGING<<" instead of "<<opts->logLevel<<std::endl;
                    opts->logLevel = CRASS_DEF_MAX_LOGGING;
                }
                break;
            case 'L':
                opts->longDescription = true;
                break;
            case 'n':
                from_string<unsigned int>(opts->minNumRepeats, optarg, std::dec);
                if (opts->minNumRepeats < 2) 
                {
                    std::cerr<<PACKAGE_NAME<<" [ERROR]: The mininum number of repeats cannot be less than 2"<<std::endl;
                    usage();
                    exit(1);
                }
                break;
            case 'o': 
                opts->output_fastq = optarg; 
                // just in case the user put '.' or '..' or '~' as the output directory
                if (opts->output_fastq[opts->output_fastq.length() - 1] != '/')
                {
                    opts->output_fastq += '/';
                }
                
                // check if our output folder exists
                struct stat file_stats;
                if (0 != stat(opts->output_fastq.c_str(),&file_stats)) 
                {
                    RecursiveMkdir(opts->output_fastq);
                }
                // check that the directory is writable
                else if(access(optarg, W_OK))
                {
                    std::cerr<<PACKAGE_NAME<<" [ERROR]: You do not have permission to write to "<<optarg<<std::endl;
                    exit(1);
                }
                break;
            case 'r': 
#ifdef RENDERING 
                opts->noRendering = true; 
#endif
                break;
            case 's': 
                from_string<unsigned int>(opts->lowSpacerSize, optarg, std::dec);
                if (opts->lowSpacerSize < 8) 
                {
                    std::cerr<<PACKAGE_NAME<<" [WARNING]: The lower bound for spacer sizes cannot be "<<opts->lowSpacerSize<<" changing to "<<CRASS_DEF_MIN_SPACER_SIZE<<std::endl;
                    opts->lowSpacerSize = CRASS_DEF_MIN_SPACER_SIZE;
                }
                break;
            case 'S': 
                from_string<unsigned int>(opts->highSpacerSize, optarg, std::dec);
                break;
            case 'V': 
                versionInfo(); 
                exit(1); 
                break;
            case 'w': 
                from_string<unsigned int>(opts->searchWindowLength, optarg, std::dec);
                if ((opts->searchWindowLength < CRASS_DEF_MIN_SEARCH_WINDOW_LENGTH) || (opts->searchWindowLength > CRASS_DEF_MAX_SEARCH_WINDOW_LENGTH))
                {
                    std::cerr<<PACKAGE_NAME<<" [WARNING]: Specified window length higher than max. Changing window length to " << CRASS_DEF_OPTIMAL_SEARCH_WINDOW_LENGTH << " instead of " << opts->searchWindowLength<<std::endl;
                    // Change window length
                    opts->searchWindowLength = CRASS_DEF_OPTIMAL_SEARCH_WINDOW_LENGTH;
                }
                break;        
            case 0:
#ifdef SEARCH_SINGLETON
                if (strcmp("searchChecker", long_options[index].name) == 0) opts->searchChecker = optarg;
#endif
                break;
            default: 
                versionInfo(); 
                usage(); 
                exit(1); 
                break;
        }
    }
    // Sanity checks for the high and low dr size
    if (opts->lowDRsize >= opts->highDRsize) 
    {
        std::cerr<<PACKAGE_NAME<<" [ERROR]: The lower direct repeat bound is bigger than the higher bound ("<<opts->lowDRsize<<" >= "<<opts->highDRsize<<")"<<std::endl;
        usage();
        exit(1);
    }
    // Sanity checks for the high and low spacer size
    if (opts->lowSpacerSize >= opts->highSpacerSize) 
    {
        std::cerr<<PACKAGE_NAME<<" [ERROR]: The lower spacer bound is bigger than the higher bound ("<<opts->lowSpacerSize<<" >= "<<opts->highSpacerSize<<")"<<std::endl;
        usage();
        exit(1);
    }
    
    
    
    return optind;
}



//**************************************
// rock and roll
//**************************************
int main(int argc, char *argv[]) 
{
    if(argc == 1) 
    {
        usage();
        exit(1);
    }
#ifdef PERFORM_CRASS_ASSEMBLY
   
    if (strcmp(argv[1], "assemble") == 0) 
    {
        // our user wants to do an assembly so let's load up the assembler main function
        //return assemblyMain(argc - 1 , argv + 1);
        std::cerr<< "As of version 0.3 of "<<PACKAGE_NAME<<" the assembly wrapper is its own executable.  Please use "<<PACKAGE_NAME<<"-assember now"<<std::endl;
        return EXIT_FAILURE;
    }
#endif
    /* application of default options */
    options opts;
    opts.logLevel              = CRASS_DEF_DEFAULT_LOGGING;              // level of verbosity allowed in the log file
    opts.reportStats           = CRASS_DEF_STATS_REPORT;                 // print a starts report currently not used
    opts.lowDRsize             = CRASS_DEF_MIN_DR_SIZE;                  // the lower size limit for a direct repeat
    opts.highDRsize            = CRASS_DEF_MAX_DR_SIZE;                  // the upper size limit for a direct repeat
    opts.lowSpacerSize         = CRASS_DEF_MIN_SPACER_SIZE;              // the lower limit for a spacer
    opts.highSpacerSize        = CRASS_DEF_MAX_SPACER_SIZE;              // the upper size limit for a spacer
    opts.output_fastq          = CRASS_DEF_OUTPUT_DIR;                   // the output directory for the output files
    opts.delim                 = CRASS_DEF_STATS_REPORT_DELIM;           // delimiter used in stats report currently not used
    opts.kmer_clust_size       = CRASS_DEF_K_CLUST_MIN;                  // number of kmers needed to be shared to add to a cluser
    opts.searchWindowLength    = CRASS_DEF_OPTIMAL_SEARCH_WINDOW_LENGTH; // option 'w'used in long read search only
    opts.minNumRepeats         = CRASS_DEF_DEFAULT_MIN_NUM_REPEATS;      // option 'n'used in long read search only
    opts.logToScreen           = CRASS_DEF_LOGTOSCREEN;                  // log to std::cout rather than to the log file
    opts.coverageBins          = CRASS_DEF_NUM_OF_BINS;                  // The number of bins of colours
    opts.graphColourType       = CRASS_DEF_GRAPH_COLOUR;                 // the colour type of the graph
    opts.longDescription       = CRASS_DEF_SPACER_LONG_DESC;             // print a long description for the final spacer graph
    opts.showSingles           = CRASS_DEF_SPACER_SHOW_SINGLES;          // print singletons when making the spacer graph
    opts.cNodeKmerLength       = CRASS_DEF_NODE_KMER_SIZE;               // length of the kmers making up a crisprnode
#ifdef DEBUG
    opts.noDebugGraph          = false;                                  // Even if DEBUG preprocessor macro is set do not produce debug graph files
#endif
#ifdef SEARCH_SINGLETON
    opts.searchChecker        = "";                                     // Name of file containing 
#endif
#ifdef RENDERING
    opts.layoutAlgorithm       = DEFAULT_RENDERING_ALGORITHM;            // the graphviz layout algorithm to use
    opts.noRendering           = false;                                  // Even if RENDERING preprocessor macro is set do not produce any rendered images
#else
    opts.layoutAlgorithm       = "unset";
#endif
    opts.covCutoff             = CRASS_DEF_COVCUTOFF;

    int opt_idx = processOptions(argc, argv, &opts);

    if (opt_idx >= argc) 
    {
        std::cerr<<PACKAGE_NAME<<" [ERROR]: Specify sequence files to process!"<<std::endl;
        usage();
        exit(1);
    }
    
    // always make a timestamp
    std::string logFileName;
    time_t rawtime;
    struct tm * timeinfo;
    char buffer [80];
    
    time ( &rawtime );
    timeinfo = localtime ( &rawtime );
    
    strftime (buffer,80,"%d_%m_%Y_%H%M%S",timeinfo);
    std::string timestamp(buffer);

    // initialize the log file
    if (opts.logToScreen) 
    {
        logFileName = "";
    } 
    else 
    {
        // create a time stamp for the log file
        logFileName = opts.output_fastq + PACKAGE_NAME +"."+timestamp+".log";
        
    }

    
    intialiseGlobalLogger(logFileName, opts.logLevel);
    
    // The remaining command line arguments are FASTQ(s) to process
    // put them in a vector to pass to the workhorse
    std::vector<std::string> seq_files;
    
    while (opt_idx < argc) 
    {
        //
        std::string seq_file(argv[opt_idx]);
        
        seq_files.push_back(seq_file);
        
        opt_idx++;
    }
    logTimeStamp();
    
    std::string cmd_line;
    for (int i = 0; i < argc; ++i) {
        cmd_line += argv[i];
        cmd_line += ' ';
    }

    WorkHorse * mHorse = new WorkHorse(&opts, timestamp,cmd_line);
    try {
#if SEARCH_SINGLETON
        debugger->headerFile(opts.searchChecker);
        debugger->processHeaderFile();
#endif
/*        
        SearchCheckerList::iterator debug_iter;
        for (debug_iter  = debugger->begin(); debug_iter != debugger->end(); debug_iter++) {
            std::cout<<debug_iter->first<<std::endl;
        }
 */
        int error_code = mHorse->doWork(seq_files);
    
#ifdef SEARCH_SINGLETON
        delete debugger;
#endif
        delete mHorse;
        return error_code;
    
    } catch (crispr::exception& e) {
        std::cerr<<e.what()<<std::endl;
        delete mHorse;
#ifdef SEACH_SINGLETON
        delete debugger
#endif
        return EXIT_FAILURE;
    }
}
